generalize `show()` to be for all subtypes of `AbstractSet`

thazhemadam opened the pull request:
Resolves #40188.

Update the error message displayed when something similar to `Code` is run.

Code -
```julia
using Test
d = Dict("hello" => "world")
@test keys(d) == ["hello"]
```

Old Output -
```julia
  Expression: keys(d) == ["hello"]
   Evaluated: ["hello"] == ["hello"]
```

New Output -
```julia
  Expression: keys(d) == ["hello"]
   Evaluated: Base.KeySet{String, Dict{String, String}}(Dict("hello" => "world")) == ["hello"]
```

Signed-off-by: Anant Thazhemadam <anant.thazhemadam@gmail.com>

NHDaly reviewed (commented):
Can you include some examples of the new behavior in the PR description?

Based on the test failures, it looks like it might not be _quite_ this simple.

For example, from the Linux64 tests:
https://build.julialang.org/#/builders/69/builds/1912/steps/5/logs/stdio

```julia
Error in testset sets:
Test Failed at /buildworker/worker/tester_linux64/build/share/julia/test/sets.jl:101
  Expression: sprint(show, Set(['a'])) == "Set(['a'])"
   Evaluated: "Set{Char}(['a'])" == "Set(['a'])"
Error in testset show:
Test Failed at /buildworker/worker/tester_linux64/build/share/julia/test/show.jl:1781
  Expression: showstr(Set([[Int16(1)]])) == "Set(Vector{Int16}[[1]])"
   Evaluated: "Set{Vector{Int16}}(Vector{Int16}[[1]])" == "Set(Vector{Int16}[[1]])"
```

It looks like now you're always printing the full type of the `Set` in the else-block on line 50, whereas we previously would skip the type argument, since it will be shown in the body of the Set.

------------

Actually, after thinking about this a bit more, it seems to me that we shouldn't modify the `show` definition for `Set` on line 41 at all!

Instead, I think the main fix for this issue is the change you found on `abstractdict.jl` line 56, to _remove_ the custom printing for `KeySet`, and to leave the default printing for it, which should already print it okay!

Notice that, on master, the default printing for new custom AbstractSet types is already reasonable:

```julia
julia> struct MySet{T} <: AbstractSet{T} args::Vector{T} end

julia> show(stdout, MySet(Int[1,2]))
MySet{Int64}([1, 2])
```
```julia
julia> struct MySetWrapper{T} <: AbstractSet{T} s::Set{T} end

julia> show(MySetWrapper(Set([1,2])))
MySetWrapper{Int64}(Set([2, 1]))
```
Which comes from https://github.com/JuliaLang/julia/blob/2e08ccdb12cca46b109492b1198635759bae1e31/base/show.jl#L397-L398

And already has a decent 3-argument show:

```julia
julia> struct MySet{T} <: AbstractSet{T} args::Vector{T} end

julia> Base.iterate(s::MySet, i...) = iterate(s.args, i...)

julia> Base.length(s::MySet) = length(s.args)

julia> MySet(Int[1,2])
MySet{Int64} with 2 elements:
  1
  2

julia> show(MySet(Int[1,2]))
MySet{Int64}([1, 2])
```
Which comes from this definition, here:
https://github.com/JuliaLang/julia/blob/2e08ccdb12cca46b109492b1198635759bae1e31/base/show.jl#L153

-------

That said, the default 2-argument-show for AbstractSets doesn't do the fancy thing that it does for `Set`s, which is to only show the types once, but I guess we don't necessarily _know for sure_ that the `MySet(Int[])` form even _is_ a valid constructor for my type. For example, it _wouldn't_ be a valid constructor for `MySetWrapper` in my example above.

So I'd probably opt to simplify this change to only fixing the display for `KeySet`, and then leaving everything else unchanged. 

Interested in others' thoughts!

thazhemadam commented:
> Actually, after thinking about this a bit more, it seems to me that we shouldn't modify the show definition for Set on line 41 at all!
>
> Instead, I think the main fix for this issue is the change you found on `abstractdict.jl` line 56, to remove the custom printing for KeySet, and to leave the default printing for it, which should already print it okay!

Yes, that works just fine. Leaving the `show` method in `set.jl` as-is would also fix the build failures.

> That said, the default 2-argument-show for `AbstractSets` doesn't do the fancy thing that it does for `Sets`, which is to only show the types once, but I guess we don't necessarily know for sure that the `MySet(Int[])` form even is a valid constructor for my type. For example, it wouldn't be a valid constructor for `MySetWrapper` in my example above.

Yes, but wouldn't an error be thrown for this at the constructor level itself when you try to execute the invalid constructor? 

Thank you for the convincing and detailed review @NHDaly! I'll wait a while to see if anyone else has other ideas before making the changes recommended.

NHDaly commented:
> Yes, but wouldn't an error be thrown for this at the constructor level itself when you try to execute the invalid constructor?

Yeah, it would, but part of the goal of 2-arg-show is that by default, it always shows values as _valid julia input_. That is, you should be able to copy/paste a value from the result of `show(stdout, val)` into the julia repl, and you should get the same value.

If we changed the default printing for all `AbstractSet`s, that wouldn't necessarily be true.

**Unless** having a constructor `MySetType([values...])` is a required constructor as part of the Abstract Set interface, in which case it would be fine. But I don't think it is: https://github.com/JuliaLang/julia/issues/34677

thazhemadam committed:
dispatch `show_vector()` only when `iter` is of type `ValueIterator`

Dispatch `show_vector` only when `iter` is of type `ValueIterator`
(instead of Union{Keysey, ValueIterator}).

Signed-off-by: Anant Thazhemadam <anant.thazhemadam@gmail.com>

thazhemadam commented:
@NHDaly ping - could you please review this PR?

NHDaly reviewed (dismissed):
LGTM! :) I love that this ended up being such a simple change! 😁 

Thanks for the PR!

DilumAluthge commented:
This seems to have been stale for a while.

@NHDaly Could I recruit you to do another review of this?
